src/lib.rs

```diff
@@ -139,21 +197,39 @@ fn run_fixing_loop(
         } else {
             violations.clear();
         }
+        let mut instantiated_per_file_rules: HashMap<RuleName, Arc<dyn RuleInstancePerFile>> =
+            Default::default();
         tree_sitter_grep::run_for_slice_with_callback(
             file_contents,
             tree_sitter_grep_args.clone(),
             |capture_info| {
-                let (rule, rule_listener) =
-                    aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
-                let mut query_match_context =
-                    QueryMatchContext::new(path, file_contents, rule, config);
-                (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+                let (instantiated_rule, rule_listener_index) =
+                    aggregated_queries.get_rule_and_listener_index(capture_info.pattern_index);
+                let mut query_match_context = QueryMatchContext::new(
+                    path,
+                    file_contents,
+                    instantiated_rule,
+                    config,
+                    language,
+                );
+                instantiated_per_file_rules
+                    .entry(instantiated_rule.meta.name.clone())
+                    .or_insert_with(|| {
+                        instantiated_rule
+                            .rule_instance
+                            .instantiate_per_file(&FileRunInfo {})
```

Comment:
Right now `FileRunInfo` is just kind of a placeholder, in theory it would contain whatever could be useful for rules to be able to "configure" their per-file run state